Yosemite/Yosemite/Stores/ShipmentStore.swift

```diff
@@ -0,0 +1,90 @@
+import Foundation
+import Networking
+import Storage
+import CocoaLumberjack
+
+
+// MARK: - ShipmentStore
+//
+public class ShipmentStore: Store {
+
+    /// Shared private StorageType for use during then entire ShipmentStore sync process
+    ///
+    private lazy var sharedDerivedStorage: StorageType = {
+        return storageManager.newDerivedStorage()
+    }()
+
+    /// Registers for supported Actions.
+    ///
+    override public func registerSupportedActions(in dispatcher: Dispatcher) {
+        dispatcher.register(processor: self, for: ShipmentAction.self)
+    }
+
+    /// Receives and executes Actions.
+    ///
+    override public func onAction(_ action: Action) {
+        guard let action = action as? ShipmentAction else {
+            assertionFailure("ShipmentStore received an unsupported action")
+            return
+        }
+
+        switch action {
+        case .synchronizeShipmentTrackingData(let siteID, let orderID, let onCompletion):
+            synchronizeShipmentTrackingData(siteID: siteID, orderID: orderID, onCompletion: onCompletion)
+        }
+    }
+}
+
+
+// MARK: - Services!
+//
+private extension ShipmentStore {
+
+    func synchronizeShipmentTrackingData(siteID: Int, orderID: Int, onCompletion: @escaping (Error?) -> Void) {
+        let remote = ShipmentsRemote(network: network)
+        remote.loadShipmentTrackings(for: siteID, orderID: orderID) { [weak self] (shipmentTrackingData, error) in
+            guard let readOnlyShipmentTrackingData = shipmentTrackingData else {
+                onCompletion(error)
+                return
+            }
+
+            self?.upsertShipmentTrackingDataInBackground(siteID: siteID, orderID: orderID, readOnlyShipmentTrackingData: readOnlyShipmentTrackingData) {
+                onCompletion(nil)
+            }
+        }
+    }
+}
+
+
+// MARK: - Persistence
+//
+extension ShipmentStore {
+
+    /// Updates (OR Inserts) the specified ReadOnly ShipmentTracking Entities into the Storage Layer *in a background thread*. onCompletion will be called
+    /// on the main thread!
+    ///
+    func upsertShipmentTrackingDataInBackground(siteID: Int, orderID: Int, readOnlyShipmentTrackingData: [Networking.ShipmentTracking], onCompletion: @escaping () -> Void) {
+        let derivedStorage = sharedDerivedStorage
+        derivedStorage.perform {
+            for readOnlyTracking in readOnlyShipmentTrackingData {
+                let storageTracking = derivedStorage.loadShipmentTracking(siteID: readOnlyTracking.siteID,
```

Comment:
Here we pass three different properties of the same object as three different parameters. We could have passed the instance of `Networking.ShipmentTracking` as well, as just one parameter and leave the storage entity to decide which properties it needs to use. I have noticed also this pattern in other places in the codebase, am I was curious as well about this. On the one hand, I fell like we might be leaking a bit of the abstraction (clients of the StorageType need to know what info it needs to do its work) but on the other hand, that logic maybe does not belong in StorageType. What are your thoughts on this?

Reply:
> We could have passed the instance of `Networking.ShipmentTracking` as well, as just one parameter and leave the storage entity to decide which properties it needs to use. I have noticed also this pattern in other places in the codebase, am I was curious as well about this.

Another good question. So from an "original architecture" standpoint, we wanted the `Storage` and `Networking` frameworks to be very dumb and simple. `Yosemite` (`*Store`'s in particular) are meant to be the glue that contained all of the business logic. 

In this specific  case, `Storage` has [the helpers in`StorageType+Extensions.swift`](https://github.com/woocommerce/woocommerce-ios/blob/develop/Storage/Storage/Tools/StorageType%2BExtensions.swift) which really just plugs in values for a predicate statement. From a high level, refactoring this func to take a `ShipmentTracking` param and then also return a `ShipmentTracking` would work, but perhaps you, @mindgraffiti and I need to put our heads together a better way to handle this.

> On the one hand, I fell like we might be leaking a bit of the abstraction (clients of the StorageType need to know what info it needs to do its work) but on the other hand, that logic maybe does not belong in StorageType.

In general, we did struggle with managing the abstraction of storage. Another example of this struggle is how we use `ResultsController` in VC's. It would be great to figure out a way to manage it better without adding a lot of "architectural overhead".  For now, I think ultimately we found a balance which seemed to serve us well and allow us to ship features quickly. 😄